create video_input.py

hylee817 opened the pull request:
# Description

> :memo: Please include a summary of the change. 
>  
> * Please also include relevant motivation and context.  
> * List any dependencies that are required for this change.  

## Type of change

For a new feature or function, please create an issue first to discuss it
with us before submitting a pull request.

Note: Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation update
- [x] TensorFlow 2 migration
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] A new research paper code implementation
- [ ] Other (Specify)

## Tests

> :memo: Please describe the tests that you ran to verify your changes.
>  
> * Provide instructions so we can reproduce.  
> * Please also list any relevant details for your test configuration.  

**Test Configuration**:

## Checklist

- [x] I have signed the [Contributor License Agreement](https://github.com/tensorflow/models/wiki/Contributor-License-Agreements).
- [x] I have read [guidelines for pull request](https://github.com/tensorflow/models/wiki/Submitting-a-pull-request).
- [ ] My code follows the [coding guidelines](https://github.com/tensorflow/models/wiki/Coding-guidelines).
- [ ] I have performed a self [code review](https://github.com/tensorflow/models/wiki/Code-review) of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.

hylee817 commented:
@googlebot I signed it!

yeqingli commented on official/vision/beta/projects/yt8m/dataloaders/video_input.py:
```diff
@@ -0,0 +1,364 @@
+from official.vision.beta.dataloaders import parser
+from official.vision.beta.ops import preprocess_ops_3d
+
+# IMAGE_KEY = 'image/encoded'
+# LABEL_KEY = 'clip/label/index'
```
Remove these lines.

XericZephyr commented on official/vision/beta/projects/yt8m/dataloaders/video_input.py:
```diff
@@ -0,0 +1,354 @@
+  The TFRecords must contain SequenceExamples with the sparse in64 'labels'
+  context feature and a fixed length byte-quantized feature vector, obtained
+  from the features in 'feature_names'. The quantized features will be mapped
+  back into a range between min_quantized_value and max_quantized_value.
```
You can link https://research.google.com/youtube8m/download.html for details.

XericZephyr commented on official/vision/beta/projects/yt8m/dataloaders/video_input.py:
```diff
@@ -0,0 +1,354 @@
+    # the same segment_size.
+    uniq_start_times, seg_idxs = tf.unique(start_times,
+                                           out_idx=tf.dtypes.int64)
+    # TODO(zhengxu): Ensure the segment_sizes are all same.
```
Remove.

XericZephyr commented on official/vision/beta/projects/yt8m/dataloaders/video_input.py:
```diff
@@ -0,0 +1,354 @@
+    Processed frames. Tensor of shape
+      [batch * num_test_clips, num_frames, height, width, 3].
+  """
+  if num_test_clips > 1 and not is_training:
```
What is this for?

XericZephyr commented on official/vision/beta/projects/yt8m/dataloaders/video_input.py:
```diff
@@ -0,0 +1,354 @@
+               min_quantized_value=-2,
+               ):
+    self._num_classes = input_params.num_classes
+    # self._image_key = image_key
```
Remove

XericZephyr commented on official/vision/beta/projects/yt8m/dataloaders/video_input.py:
```diff
@@ -0,0 +1,354 @@
+               contexts,
+               features,
+               input_params: exp_cfg.DataConfig,
+               # image_key: str = IMAGE_KEY,
```
Remove.

XericZephyr commented on official/vision/beta/projects/yt8m/dataloaders/video_input.py:
```diff
@@ -0,0 +1,354 @@
+  def __init__(self, input_params: exp_cfg.DataConfig):
+    self._segment_labels = input_params.segment_labels
+    self._is_training = input_params.is_training
+    self._num_test_clips = input_params.num_test_clips
```
It is not instantly clear to me what num_test_clips is here for. Please comment here with more details.

yeqingli replied:
The num_test_clips option is probably inherit from the video_input.py. This option is for using multiple segments for classification. It is probably not useful here. Can be remove.

yeqingli commented on official/vision/beta/projects/yt8m/dataloaders/video_input.py:
```diff
@@ -0,0 +1,364 @@
+'''class YT8MFrameFeatureReader(BaseReader): --> reimplement in tf2
```
I think this file is better called yt8m_input.py.

yeqingli commented on official/vision/beta/projects/yt8m/dataloaders/video_input.py:
```diff
@@ -0,0 +1,364 @@
+  ], axis)
+
+  # Update shape.
+  new_shape = tensor.get_shape().as_list()  # A copy is being made.
```
nit: It can be just `new_shape = tensor.shape.as_list()`.

yeqingli commented on official/vision/beta/projects/yt8m/dataloaders/video_input.py:
```diff
@@ -0,0 +1,364 @@
+  # Update shape.
+  new_shape = tensor.get_shape().as_list()  # A copy is being made.
+  new_shape[axis] = new_size
+  resized.set_shape(new_shape)
```
nit: Maybe check if tf.ensure_shape is applicable.

https://www.tensorflow.org/api_docs/python/tf/ensure_shape

hylee817 commented:
modifications made based on feedback:
1. removed most of the comments pointed out.
2. num_test_clip also removed.
3. sampler function added as asked for in the google doc.

KimSSung commented:
@googlebot I signed it!

hylee817 commented:
@googlebot I consent.

KimSSung commented:
@googlebot I consent.

yeqingli commented on official/modeling/hyperparams/params_dict.py:
```diff
@@ -260,7 +260,6 @@ def _get_kvs(tokens, params_dict):
       left_k, left_v = _get_kv(stripped_tokens[0], params_dict)
       right_k, right_v = _get_kv(stripped_tokens[1], params_dict)
       return left_k, left_v, right_k, right_v
-
```
Remove this empty line change?

yeqingli commented on official/vision/beta/configs/__init__.py:
```diff
@@ -20,3 +20,4 @@
 from official.vision.beta.configs import retinanet
 from official.vision.beta.configs import semantic_segmentation
 from official.vision.beta.configs import video_classification
+from official.vision.beta.projects.yt8m.configs import yt8m
```
For this project, please move this import to yt8m_train.py. The motivation of this is to make the project folder only one-way deepened on the outside yt8m/ folder.

yeqingli commented on official/vision/beta/tasks/__init__.py:
```diff
@@ -20,3 +20,4 @@
 from official.vision.beta.tasks import retinanet
 from official.vision.beta.tasks import semantic_segmentation
 from official.vision.beta.tasks import video_classification
+from official.vision.beta.projects.yt8m import yt8m_task
```
Please move this import to yt8m_train.py.

yeqingli commented on official/vision/beta/projects/yt8m/yt8m_task.py:
```diff
@@ -0,0 +1,200 @@
+# Lint as: python3
```
Could you move this file into yt8m/tasks/ folder?

yeqingli commented on official/vision/beta/projects/yt8m/dataloaders/yt8m_input.py:
```diff
@@ -0,0 +1,349 @@
+from typing import Dict, Optional, Tuple
+from absl import logging
+import tensorflow as tf
+# import utils
```
Remove this comment?

yeqingli commented on official/vision/beta/projects/yt8m/utils.py:
```diff
@@ -0,0 +1,217 @@
+# Copyright 2016 Google Inc. All Rights Reserved.
```
Is this file only used by the dataloaders?
If that's the case, could you move it into the dataloaders/ folder?

yeqingli commented on official/vision/beta/projects/yt8m/yt8m_agg_models.py:
```diff
@@ -0,0 +1,108 @@
+# Copyright 2016 Google Inc. All Rights Reserved.
```
Could you move this file into yt8m/modeling/ folder? So that the folder structure look alike to beta/ folder.

yeqingli commented on official/vision/beta/projects/yt8m/yt8m_model.py:
```diff
@@ -0,0 +1,127 @@
+import tensorflow as tf
```
Could you move this file into yt8m/modeling/ folder?

yeqingli commented on official/vision/beta/projects/yt8m/yt8m_model_test.py:
```diff
@@ -0,0 +1,71 @@
+# Lint as: python3
```
Please move this file into modeling/ folder?

yeqingli commented on official/vision/beta/projects/yt8m/yt8m_model_utils.py:
```diff
@@ -0,0 +1,84 @@
+# Copyright 2016 Google Inc. All Rights Reserved.
```
This file would be probably better in modeling/ folder.

yeqingli commented on official/vision/beta/projects/yt8m/yt8m_flags.py:
```diff
@@ -0,0 +1,156 @@
+  #   "eval_input_path", "",
+  #   "File glob defining the evaluation dataset in tensorflow.SequenceExample "
+  #   "format. The SequenceExamples are expected to have an 'rgb' byte array "
+  #   "sequence feature as well as a 'labels' int64 context feature.")
```
Remove this?

yeqingli commented on official/vision/beta/projects/yt8m/yt8m_flags.py:
```diff
@@ -0,0 +1,156 @@
+      'config_file',
+      default=None,
+      help='YAML/JSON files which specifies overrides. The override order '
+      'follows the order of args. Note that each file '
+      'can be used as an override template to override the default parameters '
+      'specified in Python. If the same parameter is specified in both '
+      '`--config_file` and `--params_override`, `config_file` will be used '
+      'first, followed by params_override.')
+
+  flags.DEFINE_string(
+      'params_override',
+      default=None,
+      help='a YAML/JSON string or a YAML file which specifies additional '
+      'overrides over the default parameters and those specified in '
+      '`--config_file`. Note that this is supposed to be used only to override '
+      'the model parameters, but not the parameters like TPU specific flags. '
+      'One canonical use case of `--config_file` and `--params_override` is '
+      'users first define a template config file using `--config_file`, then '
+      'use `--params_override` to adjust the minimal set of tuning parameters, '
+      'for example setting up different `train_batch_size`. The final override '
+      'order of parameters: default_model_params --> params from config_file '
+      '--> params in params_override. See also the help message of '
+      '`--config_file`.')
+
+  flags.DEFINE_multi_string(
+      'gin_file', default=None, help='List of paths to the config files.')
+
+  flags.DEFINE_multi_string(
+      'gin_params',
+      default=None,
+      help='Newline separated list of Gin parameter bindings.')
+
+  flags.DEFINE_string(
+      'tpu', default=None,
+      help='The Cloud TPU to use for training. This should be either the name '
+      'used when creating the Cloud TPU, or a grpc://ip.address.of.tpu:8470 '
+      'url.')
+
+  flags.DEFINE_string(
+      'tf_data_service', default=None, help='The tf.data service address')
```
I don't think you need to duplicate these flags defined in [flags.py](https://github.com/tensorflow/models/blob/master/official/common/flags.py)?

You can just either (1) directly call it inside your define_flags (2) call it in the yt8m_train.py.

```
def define_flags():
  tfm_flags.define_flags()
  # define other flags.
```

Or

```python
# in yt8m_train.py
if __name__ == '__main__':
  tfm_flags.define_flags()
  yt8m_flags.define_flags()
  app.run(main)
```

yeqingli commented on official/vision/beta/projects/yt8m/yt8m_flags.py:
```diff
@@ -0,0 +1,156 @@
+                     "Which learning rate to start with.")
+  flags.DEFINE_float(
+      "learning_rate_decay", 0.95,
+      "Learning rate decay factor to be applied every "
+      "learning_rate_decay_examples.")
+  flags.DEFINE_float(
+      "learning_rate_decay_examples", 4000000,
+      "Multiply current learning rate by learning_rate_decay "
+      "every learning_rate_decay_examples.")
+  flags.DEFINE_integer(
+      "num_epochs", 5, "How many passes to make over the dataset before "
+      "halting training.")
+  flags.DEFINE_integer(
+      "max_steps", None,
+      "The maximum number of iterations of the training loop.")
+  flags.DEFINE_integer(
+      "export_model_steps", 1000,
+      "The period, in number of steps, with which the model "
+      "is exported for batch prediction.")
+
+  # Other flags.
+  flags.DEFINE_integer("num_readers", 8,
+                       "How many threads to use for reading input files.")
+  flags.DEFINE_string("optimizer", "AdamOptimizer",
+                      "What optimizer class to use.")
+  flags.DEFINE_float("clip_gradient_norm", 1.0, "Norm to clip gradients to.")
+  flags.DEFINE_bool(
+    "log_device_placement", False,
+    "Whether to write the device on which every op will run into the "
+    "logs on startup.")
+
+
+  # Eval flags
+  # flags.DEFINE_string(
+  #   "eval_input_path", "",
+  #   "File glob defining the evaluation dataset in tensorflow.SequenceExample "
+  #   "format. The SequenceExamples are expected to have an 'rgb' byte array "
+  #   "sequence feature as well as a 'labels' int64 context feature.")
+  flags.DEFINE_boolean("run_once", False, "Whether to run eval only once.")
+  flags.DEFINE_integer("top_k", 20, "How many predictions to output per video.")
```
We usually don't define new flags for configs. It might be better to just replace these flags as fields in the config class. That way to can remove this file altogether and the users can specify these fields through --params_override or --config_file.

KimSSung commented:
https://github.com/hylee817/models/blob/2bde248588c8388f9d6cd4d48712ff9ca52d6288/official/vision/beta/ops/preprocess_ops_3d.py#L110 

'sample_sequence' function is expected to get num_steps & stride as int values, but line108 indicates that sequence_length is cast as float, which outputs:
TypeError (Input 'y' of 'Greater' Op has type int32 that does not match type float32 of argument 'x')
Since this is an internal file that is out of our scope of modification, would there be a way around to resolve this issue?
